src/Service/Bulk/BulkCreateUsersService.php

```diff
@@ -22,297 +22,142 @@
-                $this->writeUserAssignmentCsvData(
-                    $csvPath,
-                    $lineSlugs,
-                    $prefix,
-                    $csvFilename,
-                    $csvData,
-                    $assignmentCsvData
-                );
-                $this->writeUserAssignmentAggregratedCsvData(
-                    $automateCsvPath,
-                    $csvData,
-                    $assignmentCsvData
-                );
-                $slugWiseTotalUsersArray[$lineSlugs] = array_key_exists($lineSlugs, $slugWiseTotalUsersArray)
-                    ? $slugWiseTotalUsersArray[$lineSlugs] +  $slugTotalUsersCreated
-                    : $slugTotalUsersCreated;
-            }
+    /**
+     * @param AutogeneratedUser[] $users
+     *
+     * @throws \Doctrine\ORM\ORMException
+     * @throws \Doctrine\ORM\OptimisticLockException
+     */
+    protected function persistUsers(array $users): UserCollection
+    {
+        $usernames = [];
+        foreach ($users as $user) {
+            $this->userRepository->persist($user->toEntity($this->hasher));
+            $usernames[] = $user->getName();
         }
 
-        return $slugWiseTotalUsersArray;
-    }
+        $this->userRepository->flush();
 
-    private function createUserPassword(): string
-    {
-        return substr(str_shuffle('0123456789abcdefghijklmnopqrstvwxyz'), 0, 8);
+        return $this->userRepository->findAllByCriteria(
+            (new FindUserCriteria())->addUsernameCriterion(...$usernames)
+        );
     }
 
-    /**
-     * Function to get all slug data
-     * @throws LineItemNotFoundException
-     */
-    private function getAllLineItemSlugs(): void
+    protected function persistAssignment(AssignmentCollection $assignments): void
     {
-        $lineItems = $this->lineItemRepository->findAllAsCollection()->jsonSerialize();
-        if (empty($lineItems)) {
-            throw new LineItemNotFoundException('No line items were found in database.');
+        foreach ($assignments as $assignment) {
+            $this->assignmentRepository->persist($assignment);
         }
 
-        $this->setLineItemSlugData($lineItems);
+        $this->assignmentRepository->flush();
     }
 
-    private function getLastUserAssignedToLineItems(array $lineItemSlugs): array
+    /**
+     * @param LineItem[] $lineItems
+     */
+    private function getLastUserAssignedToLineItems(array $lineItems): array
     {
+        /*
+         * TODO: replace with uuid or something similar? current solution is not consistency safe
+         * */
+
         $userNameIncrementArray = [];
 
-        foreach ($lineItemSlugs as $slugKey => $slug) {
-            $assignment = $this->assignmentRepository->findByLineItemId($slugKey);
-            $userNameIncrementArray[$slug] = self::DEFAULT_USERNAME_INCREMENT_VALUE;
+        foreach ($lineItems as $lineItem) {
+            $assignment = $this->assignmentRepository->findByLineItemId((int)$lineItem->getId());
```

Comment:
do we need to cast to int here? id should be set. if not it looks like an exceptional situation

Reply:
Have no idea. It is part of previously merged solution. I just moved it.